ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala

```diff
@@ -996,39 +996,52 @@ private final class OracleQueries(
   private[this] def pathSteps(path: JsonPath): Cord =
     path.elems.foldMap(_.fold(k => (".\"": Cord) ++ k :- '"', (_: _0.type) => "[0]"))
 
-  // I cannot believe this function exists in 2021
-  // None if literal is too long
-  // ORA-40454: path expression not a literal
-  private[this] def oraclePathEscape(readyPath: Cord): Option[Fragment] = for {
-    s <- Some(readyPath.toString)
-    if s.size <= literalStringSizeLimit
-    _ = assert(
+  // Takes a literal text value and expresses it as a single-quoted value which can be embedded
+  // within json queries.
+  private[this] def oracleJsonStringEscape(readyPath: Cord): Fragment = {
+    val s = readyPath.toString
+    assert(
       !s.startsWith("'") && !s.endsWith("'"),
       "Oracle JSON query syntax doesn't allow ' at beginning or ending",
     )
-    escaped = s.replace("'", "''")
-    if escaped.length <= literalStringSizeLimit
-  } yield Fragment const0 ("'" + escaped + "'")
+    val escaped = s.replace("'", "''")
+    Fragment.const0("'" + escaped + "'")
+  }
   // ORA-01704: string literal too long
   private[this] val literalStringSizeLimit = 4000
 
-  private[this] def oracleShortPathEscape(readyPath: Cord): Fragment =
-    oraclePathEscape(readyPath).getOrElse(
+  // I cannot believe this function exists in 2021
+  // None if literal is too long
+  // ORA-40454: path expression not a literal
+  private[this] def oracleShortPathEscape(readyPath: Cord): Fragment = {
+    val escaped = oracleJsonStringEscape(readyPath)
+    if (escaped.toString.size <= literalStringSizeLimit)
+      escaped
+    else
       throw new IllegalArgumentException(s"path too long: $readyPath")
-    )
+  }
+
+  // Builds the SQL fragment for passing a value into a JSON path expression as a named variable
+  private[this] def passingAs(literal: JsValue, name: String): Fragment = {
+    val rendered = literal match {
+      case JsNumber(n) => sql"$n"
+      case JsString(s) => {
+        if (disableContractPayloadIndexing)
+          sql"$s" // We can pass the value through via a normal query param
+        else
+          oracleJsonStringEscape(s) // We must pass the value through as a literal.
+      }
+      case JsNull | JsTrue | JsFalse | JsArray(_) | JsObject(_) => sql"$literal"
```

Comment:
If you ever end up on this code path, something has gone horribly wrong; I recommend instead inlining the usages of `existsForm` and changing the `JsString` body to follow your new rules.

Reply:
I have specialised passingAsX for number and string only now, so these cases won't be hit.